src/hypothesis/internal/charstree.py

```diff
@@ -0,0 +1,310 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis (https://github.com/DRMacIver/hypothesis)
+#
+# Most of this work is copyright (C) 2013-2015 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# https://github.com/DRMacIver/hypothesis/blob/master/CONTRIBUTING.rst for a
+# full list of people who may hold copyright, and consult the git log if you
+# need to determine who owns an individual contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import os
+import sys
+import zlib
+import bisect
+import functools
+import itertools
+import unicodedata
+
+import marshal
+from hypothesis.errors import InvalidArgument
+from hypothesis.settings import hypothesis_home_dir
+from hypothesis.internal.compat import hunichr
+
+__all__ = (
+    'ascii_tree',
+    'unicode_tree',
+    'categories',
+    'category_by_codepoint',
+    'codepoints',
+    'codepoints_for_category',
+    'filter_tree',
+    'random_codepoint',
+)
+
+
+ASCII_TREE = {}
+UNICODE_TREE = {}
+
+
+def ascii_tree():
+    """Returns tree for ASCII characters."""
+    global ASCII_TREE
+    if not ASCII_TREE:
+        ASCII_TREE = filter_tree(unicode_tree(), max_codepoint=127)
+    return ASCII_TREE
+
+
+def unicode_tree():
+    """Returns tree of Unicode characters."""
+    global UNICODE_TREE
+    if not UNICODE_TREE:
+        try:
+            UNICODE_TREE = load_tree()
+        except:
+            UNICODE_TREE = make_tree()
+            try:
+                dump_tree(UNICODE_TREE)
+            except:  # pragma: no cover
+                pass
+    return UNICODE_TREE
+
+
+def categories(tree):
+    """Returns list of all categories in specified tree."""
+    return list(tree)
+
+
+def category_by_codepoint(tree, codepoint):
+    """Returns category by specified code point in the tree."""
+    categories = list(filter_tree(
+        tree, min_codepoint=codepoint, max_codepoint=codepoint))
+    if not categories:
+        return None
+    assert len(categories) == 1
+    return categories[0]
+
+
+def codepoints(tree):
+    """Iterates over all code points in specified tree.
+
+    Code points get yielded in single stream category by category.
+
+    """
+    by_category = functools.partial(codepoints_for_category, tree)
+    return itertools.chain.from_iterable(map(by_category, categories(tree)))
+
+
+def codepoints_for_category(tree, category):
+    """Iterates over all code points of the specified category in the tree."""
+    for value in iter_values(tree[category]):
+        for cp in range(value[0], value[1] + 1):
```

Comment:
This should use hypothesis.internal.compat.hrange - in Python 2 range is a list.